test/tests/functional/pbs_hook_execjob_prologue.py

```diff
@@ -0,0 +1,216 @@
+
+        attr = {
+            'resources_available.ncpus': 1,
+            'resources_available.mem': '2gb'
+        }
+        self.server.manager(MGR_CMD_SET, NODE, attr, id=self.hostA)
+        self.server.manager(MGR_CMD_SET, NODE, attr, id=self.hostB)
+        self.server.manager(MGR_CMD_SET, NODE, attr, id=self.hostC)
+
+        attr = {'Resource_List.select': '3:ncpus=1',
+                'Resource_List.place': 'scatter',
+                'Resource_List.walltime': 30}
+        j = Job(TEST_USER, attrs=attr)
+        jid = self.server.submit(j)
+
+        self.momB.log_match("Job;%s;JOIN_JOB as node" % jid, n=100,
+                            max_attempts=10, interval=2)
+        self.momC.log_match("Job;%s;JOIN_JOB as node" % jid, n=100,
+                            max_attempts=10, interval=2)
+        self.momA.log_match("Job;%s;executed prologue hook" % jid,
+                            n=100, max_attempts=10, interval=2)
+        self.momB.log_match("Job;%s;executed prologue hook" % jid,
+                            n=100, max_attempts=10, interval=2)
+        self.momC.log_match("Job;%s;executed prologue hook" % jid,
+                            n=100, max_attempts=10, interval=2)
+
+    def test_prologue_internal_error_no_fail_action(self):
+        """
+        Test a prologue hook with an internal error and no fail_action.
+        """
+        hook_name = "prologue_exception"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "x\n")
+
+        attr = {'event': 'execjob_prologue',
+                'enabled': 'True'}
+        self.server.create_import_hook(hook_name, attr, hook_body)
+
+        attr = {'Resource_List.select': 'vnode=%s' % self.hostA,
+                'Resource_List.walltime': 30}
+        j = Job(TEST_USER, attrs=attr)
+        j.set_sleep_time(1)
+        self.server.submit(j)
+
+        self.server.expect(NODE, {'state': 'free'}, id=self.hostA, offset=1)
+
+    def test_prologue_internal_error_offline_vnodes(self):
+        """
+        Test a prologue hook with an internal error and
+        fail_action=offline_vnodes.
+        """
+        attr = {'resources_available.mem': '2gb',
+                'resources_available.ncpus': '1'}
+        self.server.create_vnodes(self.hostC, attr, 3, self.momC, delall=True,
+                                  usenatvnode=True)
+
+        hook_name = "prologue_exception"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "x\n")
+        attr = {'event': 'execjob_prologue',
+                'enabled': 'True',
+                'fail_action': 'offline_vnodes'}
+        self.server.create_import_hook(hook_name, attr, hook_body)
+
+        attr = {'Resource_List.select': 'vnode=%s[0]' % self.hostC,
+                'Resource_List.walltime': 30}
+        j = Job(TEST_USER, attrs=attr)
+        self.server.submit(j)
+
+        attr = {
```

Comment:
For consistency write this as-
attr = {'state': 'offline',
           'comment': "offlined by hook '%s' due to hook error" % hook_name }